File: Year-2022/LeetCode-August-Challenge/LongestIncreasingSubsequence.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
private:
    vector<int> dp;
    int helper(vector<int> &nums, int n){
        dp.resize(2504, -1);
        int res = 1;
        if(dp[n] != -1) return dp[n];
        for(int i = 0; i < n; ++i){
            if(nums[n] > nums[i])
                res = max(res, helper(nums, i) + 1);
        }
        return dp[n] = res;
    }
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size(), res = 0;
        for(int i = 0; i < n; ++i){
            res = max(res, helper(nums, i));
        }
        return res;
    }
};
int main(){
    vector<int> nums = {10,9,2,5,3,7,101,18};
    Solution s;
    cout<<s.lengthOfLIS(nums)<<endl;
    return 0;
}

File: Year-2022/LeetCode-August-Challenge/LongestIncreasingSubsequence.java
class Solution {
    public int lengthOfLIS(int[] arr) {
        int amax = 1;
        int[]dp = new int[arr.length];
        for(int i=0; i<arr.length; i++){
            int max = 0;
            for(int j=0; j<i; j++){
                if(arr[j] < arr[i]) max = Math.max(max, dp[j]);
            }
            dp[i] = max+1;
            amax = Math.max(amax, dp[i]);
        }    
        return amax;
    }
}
public class LongestIncreasingSubsequence {
    public static void main(String[] args) {
        int[] nums = {10,9,2,5,3,7,101,18};
        Solution s = new Solution();
        System.out.println(s.lengthOfLIS(nums));
        
    }
}
